design/jwt-verification-design.md

```diff
@@ -0,0 +1,240 @@
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: jwt-verification-proxy
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: tls-cert
+    jwtProviders:
+      - name: provider-1
+        issuer: example.com
+        audiences:
+          - audience-1
+          - audience-2
+        remoteJWKS:
+          httpURI:
+            uri: https://example.com/jwks.json
+            timeout: 1s
+          cacheDuration: 5m
+  routes:
+    - conditions:
+        - prefix: /
+      jwtProvider: provider-1
+      services:
+        - name: s1
+          port: 80
+    - conditions:
+        - prefix: /css
+      services:
+        - name: s1
+          port: 80
+```
+
+
+## Alternatives Considered
+
+### OAuth2 filter
+Envoy also has an [OAuth2 HTTP filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/oauth2_filter), which supports end-to-end OAuth2/OIDC flows.
+This provides related but separate functionality to the JWT authentication filter.
+This [excellent blog post](https://www.jpmorgan.com/technology/technology-blog/protecting-web-applications-via-envoy-oauth2-filter) shows an example of how to use the OAuth2 and JWT filters together in Envoy.
+Contour [may pursue adding support for the OAuth2 filter](https://github.com/projectcontour/contour/issues/2664), but it will be designed and implemented separately.
+
+
+### Defining JWT verification rules separately from routes
+An alternate API design is to define JWT verification rules separately from routing rules, as shown in the below YAML:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: jwt-verification-proxy
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: tls-cert
+    jwtVerificationPolicy:
+      providers:
+        - name: provider-1
+          issuer: example.com
+          audiences:
+            - audience-1
+            - audience-2
+          remoteJWKS:
+            httpURI:
+              uri: https://example.com/jwks.json
+              timeout: 1s
+            cacheDuration: 5m
+      rules:
+        - match:
+            prefix: /js
+        - match:
+            prefix: /
+          providerName: provider-1
+  routes:
+    - conditions:
+        - prefix: /
+      services:
+        - name: s1
+          port: 80
+```
+
+This design has the potential benefit of reducing duplication in the routing rules, by not requiring an entire route to be duplicated in order to express an "except" condition for JWT verification.
+It also allows the user to have direct control over the order in which JWT rules are applied, versus being subject to Contour's rules for sorting routes.
+
+However, this design results in two separate places where route-related behavior are defined, and also creates cognitive overhead for users by having two separate ways of ordering match criteria.
+
+### Options for working with HTTPProxy Inclusion
```

Comment:
Added a few different options for how this feature interacts with inclusion here, input welcome.

Reply:
hm, so the tradeoff with these seems like single jwt provider vs. root propagating the jwt requirement

would it be too complicated to do some hybrid where you can still configure multiple providers but have a setting (bool flag in the provider type for example) where you can specify one to propagate down to children (the bool would only be set to true in one provider in the list)

you could also have the opt in/specific provider field per-route to override this

might be then useful to offer the ability to configure multiple jwt requirements on a route? but could be overkill

Reply:
I've updated the doc with a revision along the lines of Sunjay's above comment.